### PR TITLE
Feature/cmrarc-752 Ingest umm-G record

### PR DESCRIPTION
**Pull request by htranho**

### CODE REVIEW

File: app/models/cmr.rb [hunk 1]
```diff
@@ -355,24 +355,45 @@ def self.get_raw_granule(concept_id)
...
+
+    if format_type == 'echo10'
+      granule_result_echo10 = get_raw_granule_results_echo10(concept_id)
+      granule_result_echo10['format_type'] = 'echo10'
+      return granule_result_echo10
+    else
+      granule_dict = granule_dict['items'][0]
+      granule_dict['format_type'] = format_type
+      granule_dict['concept-id'] = granule_dict['meta']['concept-id']
+      granule_dict['concept_id'] = granule_dict['meta']['concept-id']
```
Comment (cgokey) [inline app/models/cmr.rb:378]: Let's make consistent, either uses dashes or _ but don't include both.

File: app/models/cmr.rb [hunk 2]
```diff
@@ -355,24 +355,45 @@ def self.get_raw_granule(concept_id)
...
+      raise CmrError, "CMR returned 0 hits for #{concept_id}"
+    end
+    # Access to first element because it should have only one record (concept_id)
+    raw_format = granule_dict['items'][0]['meta']['format']
+    format_type = raw_format == 'application/vnd.nasa.cmr.umm+json' ? 'umm_json' : 'echo10'
+
+    if format_type == 'echo10'
+      granule_result_echo10 = get_raw_granule_results_echo10(concept_id)
+      granule_result_echo10['format_type'] = 'echo10'
+      return granule_result_echo10
```
Comment (cgokey) [inline app/models/cmr.rb:373]: Why does this case return a "granule_result_echo10" but the normal return is a granule_dict?

File: app/models/cmr.rb [hunk 3]
```diff
@@ -355,24 +355,45 @@ def self.get_raw_granule(concept_id)
     granule_results["Granule"]
   end
 
+  # Given a granule concept id, it will pull all revisions and return the latest granule revision.
   def self.get_raw_granule_results(concept_id)
```
Comment (cgokey) [inline app/models/cmr.rb:359]: If this does in fact return a "single" granule, I think it should be renamed to "get_raw_granule"

File: app/models/cmr.rb [hunk 4]
```diff
@@ -355,24 +355,45 @@ def self.get_raw_granule(concept_id)
...
+      granule_dict['concept-id'] = granule_dict['meta']['concept-id']
+      granule_dict['concept_id'] = granule_dict['meta']['concept-id']
+      granule_dict['revision_id'] = granule_dict['meta']['revision-id']
+      granule_dict['Granule'] = granule_dict['umm']
+    end
+    granule_dict['items'] = []
+    granule_dict
+  end
+
+  def self.get_raw_granule_results_echo10(concept_id)
```
Comment (cgokey) [inline app/models/cmr.rb:386]: Likewise if this is going to return a single granule, shouldn't have _results in the name

File: app/models/cmr.rb [hunk 5]
```diff
@@ -472,15 +490,30 @@ def self.random_granules_from_collection(collection_concept_id, granule_count =
...
-          granule_data = Cmr.granule_list_from_collection(collection_concept_id, page_num)["result"]
+          granule_data = Cmr.granule_list_from_collection(collection_concept_id, page_num)["items"][0]
         end
         granule_data_list.push(granule_data)
       end
     end
-
-    return granule_data_list
+    result_granule_list = []
+    # todo: process the list granule_data_list
```
Comment (cgokey) [inline app/models/cmr.rb:501]: remove the word "todo: "

File: app/models/cmr.rb [hunk 6]
```diff
@@ -472,15 +490,30 @@ def self.random_granules_from_collection(collection_concept_id, granule_count =
...
+      current_granule = Hash.new
+      raw_format = granule_data['meta']['format']
+      format_type = raw_format == 'application/vnd.nasa.cmr.umm+json' ? 'umm_json' : 'echo10'
+      if format_type == 'echo10'
+        current_granule = get_raw_granule_results_echo10(granule_data['meta']['concept-id'])
+      else
+        current_granule['concept-id'] = granule_data['meta']['concept-id']
+        current_granule['concept_id'] = granule_data['meta']['concept-id']
+        current_granule['revision_id'] = granule_data['meta']['revision-id']
+        gracurrent_granulenule_dict['Granule'] = granule_data['umm']
```
Comment (cgokey) [inline app/models/cmr.rb:512]: this looks like a typo.

File: test/controllers/collections_controller_test.rb [hunk 7]
```diff
@@ -124,19 +147,52 @@ class CollectionsControllerTest < ActionController::TestCase
 
       #Since a granule is chosen at random, a full mock can not be used.
       #in this instance, we return a set collection of results for any call using this concept id and granule keyword.
```
Comment (cgokey) [inline test/controllers/collections_controller_test.rb:149]: We should include 2 tests, 1) A test for including a random echo10 granule, 2) A test for including a random umm-granule.   Just to verify things work.

File: test/controllers/granules_controller_test.rb [hunk 8]
```diff
@@ -42,7 +42,40 @@ class GranulesControllerTest < ActionController::TestCase
           }).
         to_return(status: 200, body: get_stub('search_granules_by_collection_C1000000020-LANCEAMSR2.xml'), headers: {})
 
+      stub_request(:get, "#{Cmr.get_cmr_base_url}/search/granules.umm_json?collection_concept_id=C1000000020-LANCEAMSR2&page_num=1&page_size=10").
```
Comment (cgokey) [inline test/controllers/granules_controller_test.rb:45]: Likewise 2 tests, 1) for random echo10, 2) for a random umm-json

File: test/features/can_search_collection_test.rb [hunk 9]
```diff
@@ -42,14 +42,22 @@ class CanSearchCollectionTest < SystemTestCase
...
+    stub_request(:get, "#{Cmr.get_cmr_base_url}/search/granules.umm_json?collection_concept_id=C1996546695-GHRC_DAAC&page_size=10&page_num=1").
+      with(
+        headers: {
+          'Accept'=>'*/*',
+          'Accept-Encoding'=>'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
+          'User-Agent'=>'Ruby'
+        }).
+      to_return(status: 200, body: '{"hits" : 0, "took" : 105, "items" : []}', headers: {})
   end
   describe 'search cmr collection' do
```
Comment (cgokey) [inline test/features/can_search_collection_test.rb:62]: Maybe this should be called, search and ingest a collection,   And have a test for ingesting a collection with echo10 granule and another for a umm json record.

File: app/models/cmr.rb [hunk 10]
```diff
@@ -252,20 +252,19 @@ def self.format_collection(raw_collection, data_format = "echo10")
...
+    granule_data["Granule"] = flatten_format_granule_data(granule_data["Granule"], granule_data['format_type'])
     granule_data
   end
 
-  def self.format_granule_data(granule_raw_data)
+  def self.flatten_format_granule_data(granule_raw_data, format)
     results_hash = flatten_collection(granule_raw_data)
-    add_required_fields(results_hash, DesiredFields.instance.get_format_fields('echo10_granule'))
+    if format == 'echo10'
+      add_required_fields(results_hash, DesiredFields.instance.get_format_fields('echo10_granule'))
```
[the remark this answers is not in the document]
Comment (cgokey) [reply app/models/cmr.rb:264]: No verify it should be desired.

File: test/controllers/collections_controller_test.rb [hunk 11]
```diff
@@ -48,6 +48,14 @@ class CollectionsControllerTest < ActionController::TestCase
 
           }).
         to_return(status: 200, body: get_stub('search_granules_G309210-GHRC.xml'), headers: {})
+      stub_request(:get, "https://cmr.sit.earthdata.nasa.gov/search/granules.umm_json?concept_id=G309210-GHRC").
```
[the remark this answers is not in the document]
Comment (cgokey) [reply test/controllers/collections_controller_test.rb:51]: Out the scope of this ticket, will keep eventually we should clean up the tests to be consistent.

**Review by bademiluyi — APPROVED** [review]

**Review by tle1989 — APPROVED** [review]

**Pull request merged**